src/priceClient/adapters/defiLlama.ts

```diff
@@ -0,0 +1,81 @@
+import assert from "assert";
+import { PriceFeedAdapter, TokenPrice } from "../priceClient";
+import { BaseHTTPAdapter } from "./baseAdapter";
+
+type DefiLlamaTokenPrice = {
+  price: number;
+  symbol: string;
+  timestamp: number;
+  confidence: number;
+};
+
+type DefiLlamaPriceResponse = {
+  coins: {
+    [key: string]: DefiLlamaTokenPrice;
+  };
+};
+
+const defaultTimeout = 5000; // mS
+
+export class PriceFeed extends BaseHTTPAdapter implements PriceFeedAdapter {
+  private _minConfidence: number;
+
+  constructor(name = "DefiLlama", { timeout, minConfidence = 0.9 }: { timeout?: number; minConfidence?: number }) {
+    super(name, "coins.llama.fi", { timeout: timeout ?? defaultTimeout });
+    this._minConfidence = minConfidence;
+  }
+
+  get minConfidence(): number {
+    return this._minConfidence;
+  }
+
+  set minConfidence(minConfidence: number) {
+    assert(minConfidence >= 0.0 && minConfidence <= 1.0);
+    this._minConfidence = minConfidence;
+  }
+
+  async getPriceByAddress(address: string, currency = "usd"): Promise<TokenPrice> {
+    const price = await this.getPricesByAddress([address], currency);
+    return price[0];
+  }
+
+  async getPricesByAddress(addresses: string[], currency = "usd"): Promise<TokenPrice[]> {
+    if (currency != "usd") throw new Error(`Currency ${currency} not supported by DefiLlama`);
+
+    const path = "prices/current/" + addresses.map((address) => `ethereum:${address}`).join();
+    const tokenPrices: unknown = await this.query(path, {});
+    if (!this.validateResponse(tokenPrices))
+      throw new Error(`Unexpected ${this.name} response: ${JSON.stringify(tokenPrices)}`);
+
+    return addresses.map((addr: string) => {
+      const tokenPrice = tokenPrices.coins[`ethereum:${addr.toLowerCase()}`];
```

Comment:
I could be wrong, but right here it looks like you assume that the addresses returned are lower-case, but in the validation function you allow for uppercase characters in the address.

Reply:
You're spot on - it wasn't correct.

I've done a quick test with DefiLlama - they respect whatever format the address is provided in, and seem to respond with the same address. Therefore, the regex has been updated to accept all possible cases (because it doesn't know what the request was), and the subsequent lookup is: `tokenPrices.coins[\`ethereum:${addr}\`]`.